### ISSUE TITLE
A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x290... while deleting object from Realm.

### ISSUE THREAD

**Issue by Mateuszg7** [realm/realm-java#1982]
I have A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x290 in tid 18295 (app) while deleting object from RealmList in RecyclerView. It doesn't matter if I do this async or not.

**Comment by NikoYuwono** [comment]
What version of realm are you using? And what function did you call?

**Comment by cmelchior** [comment]
Hi @Mateuszg7 
Please see https://github.com/realm/realm-java/blob/master/CONTRIBUTING.md#filing-issues when creating bugs for us to effectively help you.

**Comment by Mateuszg7** [comment]
I was using Realm version 0.84.1 and I checked it too on 0.86.
I tried two methods:
1. Some RealmList<Dog> dogs...
dogs.remove(position);
2, dogs.get(position).removeFromRealm();
Of course all in transaction.

**Comment by kneth** [comment]
@Mateuszg7 Can you share a code snippet?

**Comment by michbas796** [comment]
I have found workaround. Code snippet from my application:

realm.removeAllChangeListeners();
realm.beginTransaction();
listItems.remove(position); 
realm.commitTransaction();      
adapter.notifyItemRemoved(position);

In my case SIGSEGV occures if I switch to another activity, then return to previous and remove item from RealmResults. Before displaying the second activity item is removed correctly. 

I use RealmRecyclerView and RealmBasedRecyclerViewAdapter from https://github.com/thorbenprimke/realm-recyclerview. Realm version: 0.86.1

**Comment by kneth** [comment]
@michbas796 I noticed that you mention `RealmCyclerView` so I have created an issue there in order to get some insight on the interaction between the two components.

**Comment by thorbenprimke** [comment]
Yea, the workout isn't good. You should be able to just remove from the realm and the RealmRecyclerView will update automatically (via change listener) if it's enabled. 

@michbas796, I generally delete directly from the Realm (via re-querying the item by id and then calling `removeFromRealm`) but deleting from the `realmResults` directly should work as well. 

@kneth, I'm calling `realmResults.size()` in a few places in the adapter when the `RealmChangeListener` is triggered. Not sure if that could be the cause based on #1945.

@michbas796, since you mentioned this only happens after switching between activities and that removing the change listener resolves it, there could be a bug there (on the `RealmRecyclerView` side). 
Any way you could share your screen/activity class(es) via gist?

**Comment by michbas796** [comment]
Thanks for reply. 

@thorbenprimke 
My code is public: https://github.com/michbas796/PUM/tree/master/app/src/main/java/pl/polsl/pum2/shoppingapp/gui

Mentioned activities are MainActivity and ShoppingListActivity. I switch to ShoppingListActivity is from AllShoppingListsFragment. Sorry for Polish commit descriptions. This is just simple university project (not completed yet) created to pass mobile programming course.

Another issue in RealmRecyclerView or RealmBasedRecyclerViewAdapter is that when I use async query (findAllAsync method) recycler view is empty after swich to other fragment and return to previous. It is populated again only after recreating activity for example by rotating screen.

**Comment by kneth** [comment]
@maciejwitowski From your repository it looks like you have upgraded to Realm 0.87.1. Both 0.86.1 and 0.87.1 had a few fixes for async.

Can you give us a clear path on how to obtain the crash?

**Comment by maciejwitowski** [comment]
@kneth I think you meant @michbas796 , not me :)

**Comment by thorbenprimke** [comment]
@kneth - I can repro the crash when navigating between the two screen, `AllShoppingListsFragment` and `ShoppingListFragment`, and do a delete once the `AllShoppingListsFragment` is displayed again. 
I tracked it down to the `RealmBasedRecyclerViewAdapter` subscribing to change events but not unsubscribing. Then when the delete happens the event is received for a RealmResult that is not longer valid (so I assume). Thus the `RealmBasedRecyclerViewAdapter` needs a method that can be called to clean things up when an activity/fragment is destroyed/detached.

**Comment by kneth** [comment]
@maciejwitowski Damn autocompletion - sorry.

**Comment by kneth** [comment]
@thorbenprimke Thanks for the update.

**Comment by beeender** [comment]
I checked the code https://github.com/michbas796/PUM/tree/master/app/src/main/java/pl/polsl/pum2/shoppingapp/gui , I think it would be the same issue with #1945 

```
ShoppingList shoppingList = realm.where(ShoppingList.class).equalTo("name", listName).findFirst();
RealmResults<ShoppingListItem> itemsWithPrice = shoppingList.getItems().where().greaterThan("price", 0.0).equalTo("isBought", true).findAll();
```
The `RealmResults` is built on another `RealmResults`.

But let keep this issue open until we verify it.

**Comment by michbas796** [comment]
I have changed code in my repository  https://github.com/michbas796/PUM/tree/master/app/src/main/java/pl/polsl/pum2/shoppingapp/gui. If you still want to use old version for testing please switch to Test branch.

**Comment by thorbenprimke** [comment]
@michbas796 - I have added a `close` method to the `RealmBasedRecyclerViewAdapter`. If you add this to your `onStop` methods before the `realm.close()` it will fix the crash. 
Once that's done, you can change your `removeItem` method back to 

```
        realm.beginTransaction();
        listItems.remove(position);
        realm.commitTransaction();
```
Also you need to update the dependency to
```
compile 'com.github.thorbenprimke:realm-recyclerview:0.9.13'
```

**Comment by michbas796** [comment]
@thorbenprimke 
Thanks.

**Comment by kneth** [comment]
@thorbenprimke Thanks for the fix.
@michbas796 I'll close the issue. Don't hesitate to reopen it or create a new issue if you experience problems.
